src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs

```diff
@@ -59,8 +59,9 @@ public override void AdvanceTo(SequencePosition consumed)
 
         public override void AdvanceTo(SequencePosition consumed, SequencePosition examined)
         {
-            OnAdvance(_readResult, consumed, examined);
+            // Ensure we consume data from the RequestBodyPipe before sending WINDOW_UPDATES to the client.
             _context.RequestBodyPipe.Reader.AdvanceTo(consumed, examined);
+            OnAdvance(_readResult, consumed, examined);
```

Comment:
Are consumed and examined still referencing valid objects after calling AdvanceTo?

Reply:
Good point. I think this works today but is not guaranteed to in the future. I'll rewrite to examine the SequencePositions before calling AdvanceTo(), but still send the WINDOW_UPATE afterwards.